vvc_deblock.c: add checkasm cases for strong, one-side, weak, mixed, shift

stone-d-chen opened the pull request:
Add cases by iteratively setting max_len_{p, q} and shift.

randomize_chroma_buffers() now takes both buffers and a parameter gen_strong. It will also internally handle buffer copying.

stone-d-chen commented:
New tests

Note: Bench numbers aren't accurate.

```
AVX:
 - vvc_deblock.chroma [OK]
checkasm: all 24 tests passed
vvc_h_loop_filter_chroma_8_mix_no-shift_c: 60.1
vvc_h_loop_filter_chroma_8_mix_no-shift_avx: 80.1
vvc_h_loop_filter_chroma_8_mix_shift_c: 100.1
vvc_h_loop_filter_chroma_8_mix_shift_avx: 70.1
vvc_h_loop_filter_chroma_8_one-side_no-shift_c: 90.1
vvc_h_loop_filter_chroma_8_one-side_no-shift_avx: 70.1
vvc_h_loop_filter_chroma_8_one-side_shift_c: 139.8
vvc_h_loop_filter_chroma_8_one-side_shift_avx: 70.1
vvc_h_loop_filter_chroma_8_strong_no-shift_c: 80.1
vvc_h_loop_filter_chroma_8_strong_no-shift_avx: 70.1
vvc_h_loop_filter_chroma_8_strong_shift_c: 140.1
vvc_h_loop_filter_chroma_8_strong_shift_avx: 70.1
vvc_h_loop_filter_chroma_8_weak_no-shift_c: 60.1
vvc_h_loop_filter_chroma_8_weak_no-shift_avx: 20.1
vvc_h_loop_filter_chroma_8_weak_shift_c: 89.8
vvc_h_loop_filter_chroma_8_weak_shift_avx: 20.1
vvc_h_loop_filter_chroma_10_mix_no-shift_c: 59.8
vvc_h_loop_filter_chroma_10_mix_no-shift_avx: 30.1
vvc_h_loop_filter_chroma_10_mix_shift_c: 110.1
vvc_h_loop_filter_chroma_10_mix_shift_avx: 30.1
vvc_h_loop_filter_chroma_10_one-side_no-shift_c: 89.8
vvc_h_loop_filter_chroma_10_one-side_no-shift_avx: 30.1
vvc_h_loop_filter_chroma_10_one-side_shift_c: 140.1
vvc_h_loop_filter_chroma_10_one-side_shift_avx: 30.1
vvc_h_loop_filter_chroma_10_strong_no-shift_c: 90.1
vvc_h_loop_filter_chroma_10_strong_no-shift_avx: 30.1
vvc_h_loop_filter_chroma_10_strong_shift_c: 140.1
vvc_h_loop_filter_chroma_10_strong_shift_avx: 30.1
vvc_h_loop_filter_chroma_10_weak_no-shift_c: 60.1
vvc_h_loop_filter_chroma_10_weak_no-shift_avx: 30.1
vvc_h_loop_filter_chroma_10_weak_shift_c: 80.1
vvc_h_loop_filter_chroma_10_weak_shift_avx: 20.1
vvc_h_loop_filter_chroma_12_mix_no-shift_c: 70.1
vvc_h_loop_filter_chroma_12_mix_no-shift_avx: 29.8
vvc_h_loop_filter_chroma_12_mix_shift_c: 99.8
vvc_h_loop_filter_chroma_12_mix_shift_avx: 70.1
vvc_h_loop_filter_chroma_12_one-side_no-shift_c: 90.1
vvc_h_loop_filter_chroma_12_one-side_no-shift_avx: 70.1
vvc_h_loop_filter_chroma_12_one-side_shift_c: 130.1
vvc_h_loop_filter_chroma_12_one-side_shift_avx: 70.1
vvc_h_loop_filter_chroma_12_strong_no-shift_c: 90.1
vvc_h_loop_filter_chroma_12_strong_no-shift_avx: 70.1
vvc_h_loop_filter_chroma_12_strong_shift_c: 139.8
vvc_h_loop_filter_chroma_12_strong_shift_avx: 70.1
vvc_h_loop_filter_chroma_12_weak_no-shift_c: 50.1
vvc_h_loop_filter_chroma_12_weak_no-shift_avx: 30.1
vvc_h_loop_filter_chroma_12_weak_shift_c: 80.1
vvc_h_loop_filter_chroma_12_weak_shift_avx: 20.1
```

nuomi2021 commented on tests/checkasm/vvc_deblock.c:
```diff
@@ -72,59 +72,67 @@ static void randomize_params(int32_t beta[4], int32_t tc[4], const int size)
 
-static void randomize_chroma_buffers(int type, int beta[4], int32_t tc[4],
-   uint8_t *buf, ptrdiff_t xstride, ptrdiff_t ystride, const int shift, const int bit_depth)
+static void randomize_chroma_buffers(int type, int beta[4], int32_t tc[4], uint8_t *buf0, uint8_t *buf1, 
```
the type will indicate strong/weak/oneside

nuomi2021 commented on tests/checkasm/vvc_deblock.c:
```diff
@@ -134,22 +142,50 @@ static void check_deblock_chroma(const VVCDSPContext *h, int bit_depth)
-        randomize_buffers(buf0, buf1, BUF_SIZE);
-        randomize_chroma_buffers(0, beta, tc, buf, xstride, ystride, shift, bit_depth);
-        memcpy(buf0, buf1, BUF_SIZE);
+    for(int j = 0; j < sizeof(types)/sizeof(char*); ++j) {
```
this can be moved to randomize_chroma_buffers

stone-d-chen commented:
Made the suggested changes and pushed.

I also changed the function signature so that the parameters important for testing different code branches are first for clarity.

nuomi2021 commented on tests/checkasm/vvc_deblock.c:
```diff
@@ -72,59 +72,96 @@ static void randomize_params(int32_t beta[4], int32_t tc[4], const int size)
+    switch (type) {
+        case 0: // strong
+            for(int i = 0; i < 4; ++i) {
+                max_len_p[i] = 3;
```
the size of max_len_p is not always 4, if shift is zero, max_len_{p, q} is an array with 2 components. 
so you can move this to L188

nuomi2021 commented on tests/checkasm/vvc_deblock.c:
```diff
@@ -72,59 +72,96 @@ static void randomize_params(int32_t beta[4], int32_t tc[4], const int size)
-static void randomize_chroma_buffers(int type, int beta[4], int32_t tc[4],
-   uint8_t *buf, ptrdiff_t xstride, ptrdiff_t ystride, const int shift, const int bit_depth)
+static void randomize_chroma_buffers(const int bit_depth, const int type, const int shift, const int spatial_strong,
+uint8_t* max_len_p, uint8_t* max_len_q, int beta[4], int32_t tc[4], 
```
indent, better change to max_len_{p,q}[4] like beta 4.
better change beta to int32_t

nuomi2021 replied:
type is the most important thing for the entire function, maybe we can move it to the first one

nuomi2021 commented on tests/checkasm/vvc_deblock.c:
```diff
@@ -72,59 +72,96 @@ static void randomize_params(int32_t beta[4], int32_t tc[4], const int size)
+
+    uint8_t* buf = buf0 + xstride * 5;
+
+    if (spatial_strong) {
```
why we not use the type to decide strong, weak, and one side strong?

stone-d-chen replied:
It's possible to have max_len_{p,q} = 1 but the spatial activity calculation allow for strong. In this case the code will ignore the spatial activity and still use weak chroma deblocking. Vice versa, it's possible to have max_len_{p,q} = 3 but the spatial activity decide to go for weak. Essentially it allows for testing the spatial activity decision process from other elements.

nuomi2021 replied:
How about this:
```c
enum {
    DEBLOCK_LONG
    DEBLOCK_STRONG
    DEBLOCK_ONE_SIDE
    DEBLOCK_WEAK
} DEBLOCK_TYPE

struct Testcase {
    DEBLOCK_TYPE type;
    uint8_t max_q;
    uint8_t max_q;
};
```
then we can generate an array of Testcase case[]

stone-d-chen replied:
The combination would be between types, and spatial strength, types are equivalent to max_len_{p,q}.
So it would be the following:

```c
enum {
    DEBLOCK_STRONG
    DEBLOCK_ONE_SIDE
    DEBLOCK_WEAK
} DEBLOCK_TYPE

enum {
    SPATIAL_STRONG
    SPATIAL_WEAK
} SPATIAL_ACTIVITY

struct Testcase {
    DEBLOCK_TYPE type;
    SPATIAL_ACTIVITY activity;
    max_len_{p,q}[4] // assuming this is an out variable
};
```

The intent is clearer but it might be better to hold off until more checkasm tests are done on creating the structs since we'd have a second loop to pack the test cases and the current one would unpack, I'm not sure it buys us anything atm. Also there's a few very specific edge cases I think would be good to introduce, so we might want an expanded struct.

nuomi2021 commented on tests/checkasm/vvc_deblock.c:
```diff
@@ -72,59 +72,96 @@ static void randomize_params(int32_t beta[4], int32_t tc[4], const int size)
+
+    if (spatial_strong) {
+        for (int j = 0; j < size; j++) {
+            const int tc25 = TC25(j) << (bit_depth - 10);
```
weak, strong, one side strong can be mixed.
maybe we need to add some commentslike hevc_deblock.c

stone-d-chen commented:
Quick update:

- submitted a WIP branch for some perf optimizations, mainly so I didn't lose the train of thought for it; pausing now
- For some reason, the random TC / Beta values (ala HEVC) is not causing Strong spatial values to be generated, only weak; my original hacked version for testing strong/onesided used unrealistically large TC/Beta values to guarantee that strong would be calculated. **EDIT: This is fixed with the latest version, simple memcopy problem.**

nuomi2021 commented:
> be calculated. **EDIT: This is fixed with the latest version, simple memcopy problem.**
you can send a pr for this. small patches are easy to review and merge.
thank you.

stone-d-chen commented:
> you can send a pr for this. small patches are easy to review and merge.
> thank you.

Sounds good, do you mean send a patch to the mailing list?

nuomi2021 commented:
> > you can send a pr for this. small patches are easy to review and merge.
> > thank you.
> 
> Sounds good, do you mean send a patch to the mailing list?

to this repo, :)
still need some work to do before we send the patch upstream.

stone-d-chen committed:
vvc_deblock.c: add cases for strong, one-side, weak, mixed, shift

Add cases by iteratively setting max_len_{p, q}, shift, and spatial activity generation.

randomize_chroma_buffers(..) now takes both buffers, int spatial_strong, max_len_{p,q}. It will also internally handle buffer copying.

The function signature has been re-arranged so that the parameters for changing the randomization type are first.

Note: Setting the type (max_len_{p,q}) and spatial activity generation are separate parameters since it's possible for e.g. max_len_{p,q} = 1 but the spatial activity would allow for strong deblocking. That is, the tests the spatial activity calculation separately from the deblocking lengths.

vvc_deblock.c: fix spatial generation; assorted other bugs

Main bug preventing spatial generation was incorrect memcopy, buf0 has the generated values and needed to be copied into buf1.

xstride should always be 16 bytes even for 8bit integers since we're allocating and randomizing a 16*16 byte array. There's no functional change, just makes the memory view visually correct.

The offset from the beginning of the array (P3) to Q0 (where the code expects a pointer to) is xstride * 4; The main functional change is that it makes one-sided strong spatial generation less likely; one-sided strong spatial generation requires that value p1 be used instead of p3; the slight offset made this more likely. I'll need to fix this regression separately.

vvc_deblock: refactor random_chroma_buffer signature

Move bit_depth since it's less important than the other parameters.

Rename j to type for more clarity.
vvc_deblock: adjust spatial generation for one-side

If max_len_p == 1, then we substitute operations on P3 and P2 with P1

stone-d-chen commented:
@nuomi2021 this should be good to merge now

nuomi2021 commented:
thank you
